DBSError implementation

vkuznet opened the pull request:
Fixes #10962

#### Status
In development

#### Description
This PR provides an example of generic `DBSError` class which will extract proper DBS Go-based server error and yield it back to Python exception message.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
We need new versions of Pycurl and DBS3 clients, corresponding cmsdist PR can be found here https://github.com/cms-sw/cmsdist/pull/8185

#### External dependencies / deployment changes
<Does it require deployment changes? Does it rely on third-party libraries?>

vkuznet committed:
Provide DBSError class to handle DBS Go-server errors

New unit test for DBSErrors class

amaltaro commented:
@vkuznet apologies for the delay here Valentin.

What is the actual goal of this `DBSError` class? Is it to hold the dbs3-client exception (HTTPError raised whenever an HTTP request is not successful) and provide utilitarian `getter` methods for it? If so, can you please elaborate what kind of information we should be able to extract from them?

I have an idea of what you are doing here and it looks good, but we better first clarify the above concerns. Thanks

amaltaro commented:
Without paying too much attention to the code, which is work in progress and I see many required changes, here goes further thoughts:
* in addition to helpers for status code, message and reason, it would be useful to have an override for the stringyfication of the error such that we can output useful content instead of the object id.
* please create a couple of real unit tests against testbed. They need to be using the `insertBulkBlock` API. A couple of test cases I would suggest is:
  * sending a block dump that has a wrong data type somewhere
  * sending a block dump for an existent block
  * and maybe sending a block dump with 2 files, where each file provides a different meta-data (e.g., != AcqEra, PrepID, ProcessingVersion)

Once we are happy with the unit tests and the new DBSError class implementation, we can even switch those unit tests back to `integration`, such that they no longer get executed by the Jenkins setup.

For the record, method names should start with lower case: https://github.com/dmwm/WMCore/blob/master/standards/.pylintrc#L97-L98

amaltaro reviewed (changes requested):
Review request came while I was typing my message above, please see comments in my previous message.

vkuznet commented:
@amaltaro , I added unit test but for testing I used normal string json output instead of querying cmsweb-testbed. To query cmsweb-testbed someone needs to supply valid x509, i.e. jenkins environment or setup should provide that. I do not know how reliably add this, and I do not know if we actually need this. In unit test I used actual DBS error which I got manually via curl call to datatiers API. I can manually query testbed with any other API, e.g. as you suggested insertBulkBlock, and only cut-and-paste the error string into unit tests and that would be sufficient. Please have a look at provided unit tests and let me know if you agree on doing that, and I can extend unit test accordingly to your suggestions.

amaltaro commented:
@vkuznet the Jenkins setup has everything that we need to communicate with central services, so x509/authentication shouldn't be a concern.

I would suggest to use the dbs3-client API as well, to make sure that whatever error reported to the end user is the same between the unit test and the DBS3Upload component itself.
Looking into the unit tests for this component, I can actually see a few tests already marked as "integration", you might want to explore those:
https://github.com/dmwm/WMCore/blob/master/test/python/WMComponent_t/DBS3Buffer_t/DBSUpload_t.py#L338

vkuznet commented:
@amaltaro , I added dbs client into unit test, but I found that DBSClient will need an additional fix to handle properly DBS Go errors (so far DBSClient code expects DBS error as dict while Go server always returns results as list data-type). I'll make necessary changes to DBSClient and will post it here when it is ready. But so far I expect Jenkins tests to fail until we'll get new DBSClient in Jenkins with changes I'll make.

vkuznet commented:
Here is corresponding DBSClient PR https://github.com/dmwm/DBSClient/pull/75 which we need for this issue. It is required since it will properly extract DBS GO server error and provide correct HTTP error which `DBSErrors` from this PR will handle.

vkuznet commented:
Here is how client will see different representation of DBSError:
```
### HTTPError exception: HTTP Error 400: DBSError Code:109 Description:DBS DB ID error for provided entity, e.g. there is no record in DB for provided value Function:dbs.bulkblocks.getPrimaryDatasetTypeID Message:4106c5ea3bcefee148a7ec69592baa5ec1f1849a240e3af15de850f8e6e5fd81 unable to find primary_ds_type_id for primary ds type='' Error: nested DBSError Code:110 Description:DBS DB insert record error Function:dbs.GetRecID Message: Error: nested DBSError Code:113 Description:DBS validation error, e.g. input parameter does not match lexicon rules Function:dbs.primarydstypes.Insert Message: Error: nested DBSError Code:115 Description:DBS decode record failure, e.g. mailformed JSON Function:dbs.DecodeValidatorError Message: Error: validation error
### HTTPError exception representation: HTTPError 400, server code 109
### extracted server code message: DBS server code 109. Please refer to https://github.com/dmwm/dbs2go/blob/master/dbs/errors.go for details
```
The first one is `str` representation of DBS HTTPError, the second is `repr` representation, and last one is what `getServerCode` provides for HTTPError.

@amaltaro , please review and make your suggestions, we have three PRs which will be required to complete issue #10962 :
- DBSClient change, https://github.com/dmwm/DBSClient/pull/75
- PycurlClient change, https://github.com/dmwm/PycurlClient/pull/12
- this PR

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +60,53 @@ class DBSReaderError(DataMgmtError):
+        except:
+            self.data = str(ex)
+
+    def Code(self):
```
As already mentioned in my previous review, method names should start with lower case, see:
https://github.com/dmwm/WMCore/blob/master/standards/.pylintrc#L97-L98

and I would suggest to explicitly call it a getter method, e.g.: `def getCode`. Same applies to the methods below.

vkuznet replied:
done

amaltaro reviewed (changes requested):
Thanks for providing these changes. I left a few comments along the code.
Once you make these changes, please rebase and squash those commits according to the guidelines.

vkuznet commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -91,9 +92,12 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
             logging.info("About to call insert block for: %s", name)
             dbsApi.insertBulkBlock(blockDump=block)
             results.put({'name': name, 'success': "uploaded"})
+        except HTTPError as ex:
```
done

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , I did my best to adjust the code following your suggestions, and it is rebased now. Unfortunately, during many commits, rebases, I included test commit into main one and now I have single commit which includes both changes to the code and tests. I'm sorry about that, it was honest mistake as I had commits in different order to source and tests and I included test by accident. Please review and give me your feedback.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -91,7 +93,9 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-        except Exception as ex:
+        except HTTPError as ex:
+            dbsError = DBSError(ex.body)
+            results.put({'name': name, 'success': 'error', 'error': dbsError.Reason()})
```
This line has to go away, because you still don't know if this request is supposed to be marked as failed or not (and the dbsError.Reason would crash here as well).

vkuznet replied:
Alan, I think the entire block here is obsolete since it relies on DBS python exception. This is why I called DBSError here to get DBSError object. If it fails to parse exception coming from DBS server it will still has original exception as string and `getReason` will return it.

What I think should be done here is rather different, something like this:
```
dbsError = DBSError(ex.body)
dbsCode = dbsError.getServerCode()
if dbsCode == XXX:
   # it is missing block
else if dbsCode == YYY:
   # it is something else
....
```
but until we define these low level errors (which I can do now if we decide on them), then we should still parse exception (or DBS Go `"error":{"reason":...}`.

Please review again to make it clear how to adjust this codebase.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -109,6 +113,10 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
                 logging.exception(msg)
```
Two lines above, we call `parseDBSException`. I would suggest to completely remove it - and remove the function as well - given that we now have an specific class for parsing the DBS server error.

vkuznet replied:
Please elaborate more as I do not understand from your message what should be removed.

amaltaro replied:
This line:
```
                reason = parseDBSException(exString)
```

and the function implementation, give that it will no longer be used.

vkuznet replied:
ok, now it is clear, I removed those lines.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +63,53 @@ class DBSReaderError(DataMgmtError):
+        Return appropriate message based on giver HTTPError server code
+        """
+        dbsCodes = "https://github.com/dmwm/dbs2go/blob/master/dbs/errors.go"
+        if hasattr(self.data, "server_code"):
```
Isn't `self.data` supposed to be a dictionary object?

Second comment on this block is, isn't the `code` key nested into an `error` dictionary? E.g.:
```
[{"error": {"code": 123, ...
```

Lastly, this method should be returning an integer (the server_code error number). Maybe we return None if we cannot find the error/code in the error message/object?

amaltaro replied:
It must have been a miscommunication then, sorry.
What `getServerCode()` method should return is a simple integer value with the error code reported by the DBS server. Regarding the url with the DBS error code definition, I suggested to actually add them as a comment in this class, such that whenever we look into the source code, we know from where they come from and what their meaning is.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +63,53 @@ class DBSReaderError(DataMgmtError):
+
+    def getServerCode(self):
+        """
+        Return appropriate message based on giver HTTPError server code
```
Please update the docstring

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, please find further comments along the code.

Together with those, please also provide a cmsdist PR - in the initial description - providing the required change to the `py3-dbs3-client.spec` spec file (under comp_gcc630 branch). Please do not merge it though.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -104,11 +108,13 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-                reason = parseDBSException(exString)
-                msg = "Error trying to process block %s through DBS. Error: %s" % (name, reason)
-                logging.exception(msg)
+                logging.exception(dbsError.getReason())
```
Please keep the message above, such that we know how exactly such messages are meant to be searched in the logs. So, I'd implement it like:
```
msg = f"Error trying to process block {name} through DBS. Reason: {dbsError.getReason()}"
logging.exception(dbsError.getReason())
results.put({'name': name, 'success': "error", 'error': msg})
```

and in the general exception below, I'd suggest something like:
```
        except Exception as ex:
            msg = f"Hit a general exception while inserting block {name. Error: {str(ex}}"
            logging.exception(msg)
            results.put({'name': name, 'success': "error", 'error': msg})
```

vkuznet replied:
done

amaltaro replied:
this comment is still pending. It was likely marked as resolved by mistake.

amaltaro reviewed (commented):
Further comments along the code.

vkuznet commented:
Here is proposed patch for DBS server errors which will provide exact details of data injection failures: https://gist.github.com/vkuznet/9b1f0dca35e4ea429d6f42e70dab367c

vkuznet commented:
@amaltaro , then your comment about `getServerCode` (https://github.com/dmwm/WMCore/pull/11173/#discussion_r1023091000) has no sense to me since now `getServerCode` is identical to `getCode` (which returns server code)? Please review code again to resolve this ambiguity.

amaltaro commented:
Valentin, this is the error message that you shared at some point, either here or in the issue ticket:
```
[{'error': {'code': 122,
            'function': 'dbs.bulkblocks.InsertBulkBlocksConcurrently',
            'message': '',
            'reason': 'json: cannot unmarshal object into Go struct field '
                      'BulkBlocks.files of type []dbs.File',
            'stacktrace': '\n'
                          'goroutine 1461870 [running]:\n'
                          'github.com/dmwm/dbs2go/dbs.Error({0xae4ae0?, '
                          '0xc000c88000?}, 0x7a, {0x0, 0x0}, {0xa1ade4, '
                          '0x2b})\n'
                          '\t'
                          '/go/src/github.com/vkuznet/dbs2go/dbs/errors.go:172 '
                          '+0x99\n'
                          'github.com/dmwm/dbs2go/dbs.(*API).InsertBulkBlocksConcurrently(0xc000426000)\n'
                          '\t'
                          '/go/src/github.com/vkuznet/dbs2go/dbs/bulkblocks2.go:483 '
                          '+0x2d0\n'
                          'github.com/dmwm/dbs2go/web.DBSPostHandler({0xae8150, '
                          '0xc000554090}, 0xc000988200, {0x9feab6, 0xa})\n'
                          '\t'
                          '/go/src/github.com/vkuznet/dbs2go/web/handlers.go:542 '
                          '+0x1274\n'
                          'github.com/dmwm/dbs2go/web.BulkBlocksHandler({0xae8150?, '
                          '0xc000554090?}, 0x454114?)\n'
                          '\t'
                          '/go/src/github.com/vkuznet/dbs2go/web/handlers.go:957 '
                          '+0x45\n'
                          'net/http.HandlerFunc.ServeHTTP(0x1e0000c000460401?, '
                          '{0xae8150?, 0xc000554090?}, 0x0?)\n'
                          '\t/usr/local/go/src/net/http/server.go:2109 +0x2f\n'
                          'github.com/dmwm/dbs2go/web.limitMiddleware.func1({0xae8150?, '
                          '0xc000554090?}, 0x0?)\n'
                          '\t'
                          '/go/src/github.com/vkuznet/dbs2go/web/middlewares.go:111 '
                          '+0x38\n'
                          'net/http.HandlerFunc.ServeHTTP(0x94c100?, '
                          '{0xae8150?, 0xc000554090?}, 0x11?)\n'
                          '\t/usr/local/go/src/net/http/ser'},
  'exception': 400,
  'http': {'code': 400,
           'method': 'POST',
           'path': '/dbs/int/global/DBSWriter/bulkblocks',
           'remote_addr': '188.184.30.28:34563',
           'timestamp': '2022-11-11 18:52:15.941539719 +0000 UTC '
                        'm=+510288.469020244',
           'user_agent': 'curl/7.86.0',
           'x_forwarded_for': '67.249.140.245',
           'x_forwarded_host': 'cmsweb-testbed.cern.ch'},
  'message': 'DBSError Code:122 Description:DBS unable to parse JSON record '
             'Function:dbs.bulkblocks.InsertBulkBlocksConcurrently Message: '
             'Error: json: cannot unmarshal object into Go struct field '
             'BulkBlocks.files of type []dbs.File',
  'type': 'HTTPError'}]
```

your previous `getCode` implementation AND how it used to work until we started this discussion, is that the code that we were returning was ALWAYS the `400` one (Bad Request). As I explained in our Zoom call, if the error code is always the same, then we cannot really *standardize* error codes in the DBSServer and the clients.

Hence, what I suggested was the following:
* `getCode`: this is supposed to return the `http/code`, which I understand it to be added by dbs3-client and to be always `400` (Bad Request).
* `getServerCode`: is actually supposed to provide the DBS server error code - the whole standardization discussion here! - and in this case it should actually return `error/code` (one of those in the range that you created 100-1xx).

So, WMCore can remain using the `http/code` to see whether our HTTP request succeeded or not. However, in order to identify what the actual error was in the server, which caused the HTTP block injection call to fail, we need to consume the `error/code` instead.

I hope this clarifies all this discussion. If not, please let's save our time and discuss it again over Zoom.

vkuznet commented:
@amaltaro , I removed `getServerCode` as it is identical to `getCode`. I adjust unit tests and squashed the changes. Please review and provide your feedback again. If I misunderstood what you meant by `getServerCode` please provide appropriate code snippet how it is different from existing `getCode`.

vkuznet commented:
ok, now it is clear, the actual issue was that `getCode` was extracted from "error" block while it should come from "http" one. I corrected this and returned back `getServerCode`. I hope now everything is aligned per our discussion. Please have a look. May be we should rename `getCode` to `getHTTPCode` which will make it clear.

amaltaro commented:
> Please have a look. May be we should rename getCode to getHTTPCode which will make it clear.

Thumbs up for this!

vkuznet commented:
ok, now we have `getHttpCode` and `getServerCode`, so things should be clear. I adjusted unit tests and squashed the changes.

vkuznet commented:
@amaltaro , I added https://gist.github.com/vkuznet/9b1f0dca35e4ea429d6f42e70dab367c to DBS server code and built new image which is now deployed as my dev instance on testbed. Would you like me to update DBS writer with new error codes on DBS server side? This will allow to start testing them using testbed. Please let me know and I can easily update DBS writer to new version which will provide new codes.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +63,51 @@ class DBSReaderError(DataMgmtError):
+        """
+        if isinstance(self.data, dict):
+            return self.data['http']['code']
+        return self.data
```
Here, I think we should not return the whole exception if `http/code` cannot be found. I would either return None, or maybe we could return the integer 0 (but None is likely a better choice).

vkuznet replied:
return type should be consistent and I'll return 0

amaltaro replied:
There are pros/cons in keeping the return data consistent here. One disadvantage is that I cannot be sure if return code 0 is actually what was returned from the server or if it's a default in case the code value cannot be found in the exception.
Anyhow, I do not have any strong preference here and 0 is good as well.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +63,51 @@ class DBSReaderError(DataMgmtError):
+        """
+        if isinstance(self.data, dict):
+            return self.data['error']['code']
+        return self.data
```
Same comment here.

vkuznet replied:
return type should be consistent, and I'll return it 0.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +63,51 @@ class DBSReaderError(DataMgmtError):
+        """
+        if isinstance(self.data, dict):
+            return self.data['error']['message']
+        return self.data
```
And here

vkuznet replied:
the return type should be consistent, so I'll change it to empty string then

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -58,3 +63,51 @@ class DBSReaderError(DataMgmtError):
+        """
+        if isinstance(self.data, dict):
+            return self.data['error']['reason']
+        return self.data
```
And here

amaltaro commented on test/python/WMCore_t/Services_t/DBS_t/DBSErrors_t.py:
```diff
@@ -0,0 +1,107 @@
+
+    def testDBSErrorViaDBSClient(self):
+        """
+        testDBSErrorViaDBSClient tests DBSError class
```
Is this unit test failing because we need to update dbs3-client?

vkuznet replied:
the unit tests are failing because we need new dbs3-client, there is no way around this.

amaltaro reviewed (changes requested):
Valentin, please find further follow up comments along the code.

vkuznet commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -91,7 +93,9 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-        except Exception as ex:
+        except HTTPError as ex:
+            dbsError = DBSError(ex.body)
+            results.put({'name': name, 'success': 'error', 'error': dbsError.getReason()})
```
Alan, once again the if statements are artifacts of handling previous DBS python exceptions and in my view all of them should be gone.  And I doubt we should remove `results.put` to make code backward compatible. Please have a look how I propose to deal with it now.

vkuznet commented:
please review again.

vkuznet commented:
@amaltaro , I adjusted DBSUploadPoller according to your suggestions, and I added appropriate condition to handle missing dataset parents from DBS Go server, and I added appropriate message about the switch. Please review and provide your feedback.

amaltaro reviewed (changes requested):
Valentin, please find further comments along the code.

The jenkins summary report also mentions this problem:
```
src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py
    E0001, line 113: unexpected character after line continuation character (, line 113)
```
from what I can say, the syntax is correct. But can you please double check this interactively to make sure that code gets properly executed?

amaltaro commented:
@vkuznet can you please provide a PR against the cmsdist updating py3-dbs3-client spec to the version set in the requirements.txt file? Once you create it, please update the initial description of this PR, including a link to the relevant cmsdist PR.

vkuznet commented:
Alan, I created new PR: https://github.com/cms-sw/cmsdist/pull/8185 and updated this PR description. Please let me know if you are happy with this PR now.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -91,25 +93,39 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
+            # legacy code which deal with DBS python server exceptions
+            dbsError = DBSError(ex.body)
+            reason = dbsError.getReason()
+            if 'Block %s already exists' % name in reason:
```
Valentin, can you please confirm which HTTPError error field this message (Block BLAH already exists) is supposed to be propagated? Is it `reason` or `message`?

amaltaro reviewed (commented):
Valentin, I have only one question along the code that I need your confirmation. Otherwise, it looks good to me. But we need to hold off with merging it until we hear from the T0 team.

vkuznet commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -91,25 +93,39 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-                msg = "Error trying to process block %s through DBS. Error: %s" % (name, reason)
-                logging.exception(msg)
-                logging.debug("block info: %s \n", block)
+                msg = f"Error trying to process block {name} through DBS. Reason: {dbsError.getReason()}"
```
done

vkuznet commented:
Alan, I agree to keep it open until we hear from T0 folks. Meanwhile, I asked you a question should I deploy new DBS go server which will provide fine granular error codes to testbed. Please let me know.

amaltaro commented:
> Alan, I agree to keep it open until we hear from T0 folks. Meanwhile, I asked you a question should I deploy new DBS go server which will provide fine granular error codes to testbed. Please let me know.

Yes, I think it won't hurt to run the new server code in testbed.

BTW, you might want to open a new GH issue as a follow up of this one (to start adopting the DBS server error code for error handling in DBS3Upload).

vkuznet commented:
@amaltaro , the new issue to enable DBSError codes has been created over here: https://github.com/dmwm/WMCore/issues/11365

amaltaro commented:
The T0 release has been made, so we can now merge it (and the cmsdist PR as well).
